RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,177 @@
+
+We do not propose automatic marking of functions with this decorator due to the potential backwards-compatibility break it could cause, as well as the parameters that are needed in order to allow this to happen (namely the dispatcher, which isn't in our control).
+
+### Making `torch.Tensor._make_subclass` public API
+`torch.Tensor._make_subclass` will be renamed to `torch.Tensor.make_subclass` and it will become public API. This will allow `torch.Tensor.make_subclass` correspond to `numpy.ndarray.view`, with the difference that the latter can also handle viewing as a different `dtype` and not just as a different subclass.
+
+The reason for not choosing the name `view` is that it exists on `torch.Tensor` in an unrelated context. The semantics of this function will be the same as creating a shallow copy of the object and then changing its `__class__`, which means that the underlying data pointer will remain the same.
+
+### Generic implementation of `__torch_function__`
+`torch.Tensor` will gain a generic `__torch_function__` of the following form:
+
+```python
+class Tensor:
+    def __torch_tensor__(self, func, types, args, kwargs):
+        if not all(issubclass(type(self), t) for t in types):
+            return NotImplemented
+        
+        # Defer to internal implementation
+        ret = func._implementation(*args, **kwargs)
+        if type(self) is not Tensor and isinstance(ret, Tensor):
+            ret = Tensor.make_subclass(ret, type(self))
+        return ret
+```
+
+This method matches `torch` dispatch rules, so for the most part it's possible to pretend it doesn't exist. This also has the side-effect of passing subclasses through methods, and operators (since all operators are methods).
+
+This corresponds exactly to the implementation `numpy.ndarray` gains in [[4]], except for the fact that subclasses are passed through via another internal mechanism (namely the `__array_finalize__` protocol) there, as well as the fact that we are checking subclassing against `type(self)` instead of `Tensor`. This has the side-effect of ensuring unrelated class trees are not merged, which is an inconsistency in NumPy's own design. Specifically, consider the example of two direct subclasses of `torch.Tensor`. Both will return `NotImplemented`, and therefore, the check will fail and `TypeError` will be raised.
+
+Since subclasses are checked before superclasses in `__torch_function__`, it is guaranteed that the subclass implementation will be called first. In this instance, since `type(self)` is a subclass of all types, the code will continue. Since `type(self)` is not `torch.Tensor`, a view into the original data is created and returned.
+
+This also works for all operators: `__add__`, `__getitem__` and so on since in Python these operators are just dunder methods of the corresponding class.
+
```

Comment:
Can you add an `Implementation` section? No need for a lot of research to fill it, just state what you know now. We spent a ton of time on the C++ implementation of the dispatching, so a hint as to whether it can simply be mirrored in the method version or not would be nice (perhaps @ngoldbaum has a sense here about how parallel method/function code is).

Reply:
Added.